synapse/rest/media/v1/media_storage.py

```diff
@@ -0,0 +1,198 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from twisted.internet import defer, threads
+from twisted.protocols.basic import FileSender
+
+from ._base import Responder
+
+from synapse.util.logcontext import make_deferred_yieldable
+
+import contextlib
+import os
+import logging
+import shutil
+
+logger = logging.getLogger(__name__)
+
+
+class MediaStorage(object):
+    """Responsible for storing/fetching files from local sources.
+    """
+
+    def __init__(self, local_media_directory, filepaths):
+        self.local_media_directory = local_media_directory
+        self.filepaths = filepaths
+
+    @defer.inlineCallbacks
+    def store_file(self, source, file_info):
+        """Write `source` to the on disk media store, and also any other
+        configured storage providers
+
+        Args:
+            source: A file like object that should be written
+            file_info (FileInfo): Info about the file to store
+
+        Returns:
+            Deferred[str]: the file path written to in the primary media store
+        """
+        path = self._file_info_to_path(file_info)
+        fname = os.path.join(self.local_media_directory, path)
+
+        dirname = os.path.dirname(fname)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        # Write to the main repository
+        yield make_deferred_yieldable(threads.deferToThread(
+            _write_file_synchronously, source, fname,
+        ))
+
+        defer.returnValue(fname)
+
+    @contextlib.contextmanager
+    def store_into_file(self, file_info):
+        """Context manager used to get a file like object to write into, as
+        described by file_info.
+
+        Actually yields a 3-tuple (file, fname, finish_cb), where finish_cb is a
+        function that returns a Deferred that must be waited on after the file
+        has been successfully written to.
+
+        Args:
+            file_info (FileInfo): Info about the file to store
+
+        Example:
+
+            with media_storage.store_into_file(info) as (f, fname, finish_cb):
+                # .. write into f ...
+                yield finish_cb()
+        """
+
+        path = self._file_info_to_path(file_info)
+        fname = os.path.join(self.local_media_directory, path)
+
+        dirname = os.path.dirname(fname)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        finished_called = [False]
+
+        def finish():
+            # This will be used later when we want to hit out to other storage
+            # places
+            finished_called[0] = True
+            return defer.succeed(None)
+
+        try:
+            with open(fname, "wb") as f:
+                yield f, fname, finish
+        except Exception as e:
+            try:
+                os.remove(fname)
+            except Exception:
+                pass
+            raise e
```

Comment:
s/`raise e`/`raise`/, which preserves the stacktrace from the original exception.

Reply:
AIUI, `raise` raises the last exception rather than the exception in the current block. So in this case `raise` would use any error thrown by the `os.remove`

Reply:
true. In that case please can youdo the fiddling with sys.exc_info (`raise exc_info[0], exc_info[1], exc_info[2]`)?